### DISCUSSION TITLE
How to add IP SANs to agent nodes so k3s can validate certificates for agent node?

### DISCUSSION THREAD

**Discussion by natenberenstein** [k3s-io/k3s#9887] [Q&A] [answered]
**Environmental Info**
K3s version: v1.28.7+k3s1
go version: go1.21.7

Cluster Configuration:
- 1 server (Centos 7)
- 2 agents (Centos 7)

**Describe the Bug**
When applying a manifest, I get the following error:
```
Error from server: Get "https://<agent_ip>:10250/containerLogs/default/my-depl-6c59b9f5b7-zsltc/my_depl": tls: failed to verify certificate: x509: cannot validate certificate for <agent_ip> because it doesn't contain any IP SANs
```

**Discussion**
So, I have a simple manifest and the pods will pull an image from my private docker registry.  When I apply the manifests, I get the error above, and I'm not exactly sure how to go about with this problem.  And since we are behind a corporate proxy, I'm not sure what other factors to consider, but I did set the proxies in my systems.  When I run `k3s kubectl get nodes`, I get the following:
|NAME | STATUS | ROLES | AGE | VERSION |
|-------|----------|---------|------|-----------|
|fte-a10|Ready|none|23m|v1.28.8+k3s1|
|fte-a20.foo.com|Ready|none|control-plane,master|12d|v1.28.7+k3s1|
|fte-a30-2.foo.com|Ready|none|7d6h|v1.28.8+k3s1|

It seems like the server was able to talk to the agents and add them to the cluster, but I'm not sure why the error above pops up.

**Comment by brandond** [comment]
> `cannot validate certificate for <agent_ip> because it doesn't contain any IP SANs`

The kubelet serving cert automatically includes the node IPs in the TLS SAN list. If you are getting errors about missing IP SANs, then you are talking to something other than the kubelet - I suspect there is a proxy or something else in your environment.

> since we are behind a corporate proxy, I'm not sure what other factors to consider

That would do it. Ensure that you've set the NO_PROXY environment variable to a IP mask that covers all your node IPs.

**Reply by natenberenstein** [reply]
Thanks for the reply! So I set up the NO_PROXY environment variable for all the systems, but still no luck. To be specific, I set the environment variables in my .bashrc file.  Do I need to specify the NO_PROXY env variable in other configuration files, including k3s? I'm also not too familiar with networking, so any info would be great.

**Reply by brandond** [reply]
You need to set it in one of the environment files used by K3s - /etc/default/k3s or /etc/sysconfig/k3s. Just having it in your shell doesn't help the service any.

**Reply by natenberenstein** [reply, marked as answer]
Actually, I sent up the environment variables in `/etc/systemd/system/k3s.service.env` and that worked!  Server was able to assign a pod to one of the agents.  Thanks for helping me debug the issue!

**Reply by brandond** [reply]
that one also works. Be careful though as it will be overwritten if you ever re-run the install script. The other two are safe.

**Reply by natenberenstein** [reply]
Does k3 create any directories under default or sysconfig in /etc when we run k3s?  Or do we need to manually create it?

**Reply by brandond** [reply]
it doesn't create anything there, no. Those env files don't exist by default, but are loaded by systemd if they do.
